[Fast-reboot] Change fast-reboot traffic inspection to warm-reboot way for better accuracy on test result

shlomibitton opened the pull request:
Signed-off-by: Shlomi Bitton <shlomibi@nvidia.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)


### Back port request
- [ ] 201911

### Approach
#### What is the motivation for this PR?
Currently the test traffic inspection is working on real-time, which means transmitting packets and receive them with PTF infrastructure will take time to process.
This will lead to ~0-5 seconds deviation on test result, and when fast-reboot criteria must be less than 30 seconds, 5 seconds can be a very big and unwanted deviation.
In general inaccurate results are not desired, so adopting the way warm-reboot test is inspecting the traffic loss is introduced on this PR.
The difference between current fast-reboot approach and warm-reboot approach is to constantly generate traffic and record it for a few minutes, only at the end of this time period inspect the traffic and get to a conclusion.
This way we can be as accurate as possible and get a reliable test result.

#### How did you do it?
Change the current traffic generation and inspection to the one used on warm-reboot.

#### How did you verify/test it?
Run the test several times and get the same result each time, which is the correct result.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

shlomibitton commented:
/azpw run

vaibhavhd commented:
Thanks for this enhancement. Do you have an example output to share? What were the number of packet drops seen during fast-reboot, and how long was the data plane downtime in this improved calculation?

shlomibitton commented:
> Thanks for this enhancement. Do you have an example output to share? What were the number of packet drops seen during fast-reboot, and how long was the data plane downtime in this improved calculation?

@vaibhavhd you can see the following output from a test run:

2021-08-22 11:44:42 : Pcap file dumped to /tmp/capture.pcap
2021-08-22 11:44:42 : Packet flow examine started 0:05:39.077245 after the reboot
2021-08-22 11:47:00 : Disruption between packet ID 5138 and 9977. For 26.8323 
2021-08-22 11:47:00 : Disruption between packet ID 9979 and 9981. For 0.0054 
2021-08-22 11:47:00 : Disruption between packet ID 9984 and 9986. For 0.0054 
2021-08-22 11:47:13 : Disruptions happen between 0:00:52.199294 and 0:01:19.087322 after the reboot.
2021-08-22 11:47:13 : Total incoming packets captured 31160
2021-08-22 11:48:39 : Filtered pcap dumped to /tmp/capture_filtered.pcap
2021-08-22 11:48:39 : Packet flow examine finished after 0:03:57.014033
2021-08-22 11:48:39 : Dataplane disruption lasted 26.832 seconds. 4838 packet(s) lost.
2021-08-22 11:48:39 : Total disruptions count is 3. All disruptions lasted 26.843 seconds. Total 4840 packet(s) lost
2021-08-22 11:48:39 : Wait until bgp routing is up on all devices
2021-08-22 11:48:39 : SSH thread VM=10.213.84.184: Collecting logs for attempt 0
2021-08-22 11:48:39 : SSH thread VM=10.213.84.183: Collecting logs for attempt 0
2021-08-22 11:48:39 : SSH thread VM=10.213.84.182: Collecting logs for attempt 0
2021-08-22 11:48:39 : SSH thread VM=10.213.84.181: Collecting logs for attempt 0
2021-08-22 11:48:39 : Waiting till SSH threads stop
2021-08-22 11:48:40 : SSH thread VM=10.213.84.182: Log output "show log | begin session_begins_1629632293.186867
Aug 22 11:38:30 ARISTA02T1 ConfigAgent: %SYS-6-LOGMSG_INFO: Message from admin on vty3 (10.213.84.213): session_begins_1629632293.186867
Aug 22 11:38:46 ARISTA02T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.58 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:38:46 ARISTA02T1 Rib: %BGP-5-ADJCHANGE: peer fc00::75 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:40:23 ARISTA02T1 Lag: %LAG-5-MEMBER_REMOVED: Interface Ethernet1 has left Port-Channel1
Aug 22 11:40:23 ARISTA02T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to down
Aug 22 11:40:33 ARISTA02T1 Lldp: %LLDP-5-NEIGHBOR_NEW: LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce41 added on interface Ethernet1
Aug 22 11:40:34 ARISTA02T1 Lag: %LAG-5-MEMBER_ADDED: Interface Ethernet1 has joined Port-Channel1
Aug 22 11:40:34 ARISTA02T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to up
Aug 22 11:40:38 ARISTA02T1 Lldp: %LLDP-5-NEIGHBOR_GONE: Received a shutdown frame from LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce41 on interface Ethernet1
Aug 22 11:40:41 ARISTA02T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.58 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
Aug 22 11:40:41 ARISTA02T1 Rib: %BGP-5-ADJCHANGE: peer fc00::75 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
ARISTA02T1#"
2021-08-22 11:48:40 : SSH thread VM=10.213.84.181: Log output "show log | begin session_begins_1629632293.187582
Aug 22 11:38:20 ARISTA01T1 ConfigAgent: %SYS-6-LOGMSG_INFO: Message from admin on vty3 (10.213.84.213): session_begins_1629632293.187582
Aug 22 11:38:36 ARISTA01T1 Rib: %BGP-5-ADJCHANGE: peer fc00::71 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:38:36 ARISTA01T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.56 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:40:13 ARISTA01T1 Lag: %LAG-5-MEMBER_REMOVED: Interface Ethernet1 has left Port-Channel1
Aug 22 11:40:13 ARISTA01T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to down
Aug 22 11:40:23 ARISTA01T1 Lldp: %LLDP-5-NEIGHBOR_NEW: LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce40 added on interface Ethernet1
Aug 22 11:40:24 ARISTA01T1 Lag: %LAG-5-MEMBER_ADDED: Interface Ethernet1 has joined Port-Channel1
Aug 22 11:40:24 ARISTA01T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to up
Aug 22 11:40:29 ARISTA01T1 Lldp: %LLDP-5-NEIGHBOR_GONE: Received a shutdown frame from LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce40 on interface Ethernet1
Aug 22 11:40:31 ARISTA01T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.56 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
Aug 22 11:40:31 ARISTA01T1 Rib: %BGP-5-ADJCHANGE: peer fc00::71 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
ARISTA01T1#"
2021-08-22 11:48:40 : SSH thread VM=10.213.84.181: Disconnecting from VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.182: Disconnecting from VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.183: Log output "show log | begin session_begins_1629632293.187852
Aug 22 11:38:25 ARISTA03T1 ConfigAgent: %SYS-6-LOGMSG_INFO: Message from admin on vty3 (10.213.84.213): session_begins_1629632293.187852
Aug 22 11:38:41 ARISTA03T1 Rib: %BGP-5-ADJCHANGE: peer fc00::79 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:38:41 ARISTA03T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.60 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:40:18 ARISTA03T1 Lag: %LAG-5-MEMBER_REMOVED: Interface Ethernet1 has left Port-Channel1
Aug 22 11:40:18 ARISTA03T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to down
Aug 22 11:40:28 ARISTA03T1 Lldp: %LLDP-5-NEIGHBOR_NEW: LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce42 added on interface Ethernet1
Aug 22 11:40:29 ARISTA03T1 Lag: %LAG-5-MEMBER_ADDED: Interface Ethernet1 has joined Port-Channel1
Aug 22 11:40:29 ARISTA03T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to up
Aug 22 11:40:33 ARISTA03T1 Lldp: %LLDP-5-NEIGHBOR_GONE: Received a shutdown frame from LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce42 on interface Ethernet1
Aug 22 11:40:34 ARISTA03T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.60 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
Aug 22 11:40:37 ARISTA03T1 Rib: %BGP-5-ADJCHANGE: peer fc00::79 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
ARISTA03T1#"
2021-08-22 11:48:40 : SSH thread VM=10.213.84.183: Disconnecting from VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.184: Log output "show log | begin session_begins_1629632293.188694
Aug 22 11:38:26 ARISTA04T1 ConfigAgent: %SYS-6-LOGMSG_INFO: Message from admin on vty3 (10.213.84.213): session_begins_1629632293.188694
Aug 22 11:38:42 ARISTA04T1 Rib: %BGP-5-ADJCHANGE: peer fc00::7d (AS 65100) old state Established event Closed new state Idle
Aug 22 11:38:42 ARISTA04T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.62 (AS 65100) old state Established event Closed new state Idle
Aug 22 11:40:19 ARISTA04T1 Lag: %LAG-5-MEMBER_REMOVED: Interface Ethernet1 has left Port-Channel1
Aug 22 11:40:19 ARISTA04T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to down
Aug 22 11:40:28 ARISTA04T1 Lldp: %LLDP-5-NEIGHBOR_NEW: LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce43 added on interface Ethernet1
Aug 22 11:40:29 ARISTA04T1 Lag: %LAG-5-MEMBER_ADDED: Interface Ethernet1 has joined Port-Channel1
Aug 22 11:40:29 ARISTA04T1 Ebra: %LINEPROTO-5-UPDOWN: Line protocol on Interface Port-Channel1, changed state to up
Aug 22 11:40:34 ARISTA04T1 Lldp: %LLDP-5-NEIGHBOR_GONE: Received a shutdown frame from LLDP neighbor with chassisId 506b.4bcd.2e74 and portId 506b.4b8f.ce43 on interface Ethernet1
Aug 22 11:40:34 ARISTA04T1 Rib: %BGP-5-ADJCHANGE: peer 10.0.0.62 (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
Aug 22 11:40:42 ARISTA04T1 Rib: %BGP-5-ADJCHANGE: peer fc00::7d (AS 65100) old state OpenConfirm event RecvKeepAlive new state Established
ARISTA04T1#"
2021-08-22 11:48:40 : SSH thread VM=10.213.84.184: Disconnecting from VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.182: Checking BGP GR peer status on VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.182: Finishing run()
2021-08-22 11:48:40 : SSH thread for VM 10.213.84.182 finished
2021-08-22 11:48:40 : SSH thread VM=10.213.84.181: Checking BGP GR peer status on VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.181: Finishing run()
2021-08-22 11:48:40 : SSH thread for VM 10.213.84.181 finished
2021-08-22 11:48:40 : SSH thread VM=10.213.84.183: Checking BGP GR peer status on VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.183: Finishing run()
2021-08-22 11:48:40 : SSH thread for VM 10.213.84.183 finished
2021-08-22 11:48:40 : SSH thread VM=10.213.84.184: Checking BGP GR peer status on VM
2021-08-22 11:48:40 : SSH thread VM=10.213.84.184: Finishing run()
2021-08-22 11:48:40 : SSH thread for VM 10.213.84.184 finished
2021-08-22 11:48:40 : Data plane works again. Start time: 2021-08-22 11:40:22.895953
2021-08-22 11:48:40 : 
2021-08-22 11:48:40 : ==================================================
2021-08-22 11:48:40 : Report:
2021-08-22 11:48:40 : ==================================================
2021-08-22 11:48:40 : LACP/BGP were down for (extracted from cli):
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 :     10.213.84.181 - lacp:   0.000 (0) po_events: (1) bgp v4:   0.000 (0) bgp v6:   0.000 (0)
2021-08-22 11:48:40 :     10.213.84.182 - lacp:   0.000 (0) po_events: (1) bgp v4:   0.000 (0) bgp v6:   0.000 (0)
2021-08-22 11:48:40 :     10.213.84.183 - lacp:   0.000 (0) po_events: (1) bgp v4:   0.000 (0) bgp v6:   0.000 (0)
2021-08-22 11:48:40 :     10.213.84.184 - lacp:   0.000 (0) po_events: (1) bgp v4:   0.000 (0) bgp v6:   0.000 (0)
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Extracted from VM logs:
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Extracted log info from 10.213.84.181
2021-08-22 11:48:40 :     BGP IPv4 was down (seconds) : 115
2021-08-22 11:48:40 :     BGP IPv4 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv4 was gotten up after Po was up (seconds) : 7
2021-08-22 11:48:40 :     BGP IPv6 was down (seconds) : 115
2021-08-22 11:48:40 :     BGP IPv6 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv6 was gotten up after Po was up (seconds) : 7
2021-08-22 11:48:40 :     Interface Port-Channel1 was down (times) : 1
2021-08-22 11:48:40 :     PortChannel was down (seconds) : 11
2021-08-22 11:48:40 :     PortChannel went down after bgp session was down (seconds) : 97
2021-08-22 11:48:40 :     defaultdict(<type 'list'>, {})
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Extracted log info from 10.213.84.182
2021-08-22 11:48:40 :     BGP IPv4 was down (seconds) : 115
2021-08-22 11:48:40 :     BGP IPv4 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv4 was gotten up after Po was up (seconds) : 7
2021-08-22 11:48:40 :     BGP IPv6 was down (seconds) : 115
2021-08-22 11:48:40 :     BGP IPv6 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv6 was gotten up after Po was up (seconds) : 7
2021-08-22 11:48:40 :     Interface Port-Channel1 was down (times) : 1
2021-08-22 11:48:40 :     PortChannel was down (seconds) : 11
2021-08-22 11:48:40 :     PortChannel went down after bgp session was down (seconds) : 97
2021-08-22 11:48:40 :     defaultdict(<type 'list'>, {})
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Extracted log info from 10.213.84.183
2021-08-22 11:48:40 :     BGP IPv4 was down (seconds) : 113
2021-08-22 11:48:40 :     BGP IPv4 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv4 was gotten up after Po was up (seconds) : 5
2021-08-22 11:48:40 :     BGP IPv6 was down (seconds) : 116
2021-08-22 11:48:40 :     BGP IPv6 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv6 was gotten up after Po was up (seconds) : 8
2021-08-22 11:48:40 :     Interface Port-Channel1 was down (times) : 1
2021-08-22 11:48:40 :     PortChannel was down (seconds) : 11
2021-08-22 11:48:40 :     PortChannel went down after bgp session was down (seconds) : 97
2021-08-22 11:48:40 :     defaultdict(<type 'list'>, {})
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Extracted log info from 10.213.84.184
2021-08-22 11:48:40 :     BGP IPv4 was down (seconds) : 112
2021-08-22 11:48:40 :     BGP IPv4 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv4 was gotten up after Po was up (seconds) : 5
2021-08-22 11:48:40 :     BGP IPv6 was down (seconds) : 120
2021-08-22 11:48:40 :     BGP IPv6 was down (times) : 1
2021-08-22 11:48:40 :     BGP IPv6 was gotten up after Po was up (seconds) : 13
2021-08-22 11:48:40 :     Interface Port-Channel1 was down (times) : 1
2021-08-22 11:48:40 :     PortChannel was down (seconds) : 10
2021-08-22 11:48:40 :     PortChannel went down after bgp session was down (seconds) : 97
2021-08-22 11:48:40 :     defaultdict(<type 'list'>, {})
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Summary:
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Longest downtime period was 0:00:26.832340
2021-08-22 11:48:40 : Reboot time was 0:01:19.038238
2021-08-22 11:48:40 : Expected downtime is less then 0:00:45
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : Additional info:
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : INFO:10.213.84.184:PortChannel interface state changed 1 times
2021-08-22 11:48:40 : INFO:10.213.84.181:PortChannel interface state changed 1 times
2021-08-22 11:48:40 : INFO:10.213.84.183:PortChannel interface state changed 1 times
2021-08-22 11:48:40 : INFO:10.213.84.182:PortChannel interface state changed 1 times
2021-08-22 11:48:40 : --------------------------------------------------
2021-08-22 11:48:40 : ==================================================
